Event loop group shutdown with IOUring slow on Java 25

Environment: Netty version: 4.2.9.Final

Hi, we hit [a problem](https://github.com/kroxylicious/kroxylicious/issues/3188) in an integration test suite that rapidly creates and destroys some netty based test clients, where they preferred IOUring as transport, and they are running on java 25. The tests were running out of resources and on debugging we found that the netty event loop threads were unexpectedly growing throughout our tests and eventually exhausting resources somewhere. In our tests we were closing the EventLoopGroup but not blocking on the shutdown future completing. Changing the clients to use epoll fixed our test issue. Also interesting is that our test suite running on java 21 is not affected.

With some debugging I found that the EventLoopGroups were taking a very long time to be destroyed.

Here's a reproducer demonstrating that there is a difference between how long it takes to shutdown IOUring ELGs on java 21 vs java 25. It appears that it is related to how many ELGs you are trying to destroy in parallel, as well as how many threads you have in each ELG.

https://github.com/robobario/iouring-reproducer/tree/main

There are some details in the reproducer about my environment and a flamegraph attached where the vast majority of the time is spent in `jdk.internal.misc.ScopedMemoryAccess.closeScope0(MemorySessionImpl, ScopedMemoryAccess$ScopedAccessError)`, which was caused by `io.netty.channel.uring.IoUringIoHandler.destroy()`.

Maybe this is related to #15582, another problem with IOUring shutdown taking a long time.

I've reproduced the failure using @robobario code https://github.com/robobario/iouring-reproducer/tree/main

Temurin 25
```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running EventLoopGroupShutdownTest
WARNING: A restricted method in java.lang.System has been called
WARNING: java.lang.System::loadLibrary has been called by io.netty.util.internal.NativeLibraryUtil in an unnamed module (file:/home/sam/.m2/repository/io/netty/netty-common/4.2.9.Final/netty-common-4.2.9.Final.jar)
WARNING: Use --enable-native-access=ALL-UNNAMED to avoid a warning for callers in this module
WARNING: Restricted methods will be blocked in a future release unless native access is enabled

ELG shutdown gracefully in 55921 milliseconds
ELG shutdown gracefully in 39 milliseconds



ELG shutdown gracefully in 853914 milliseconds
ELG shutdown gracefully in 854083 milliseconds
ELG shutdown gracefully in 853996 milliseconds
ELG shutdown gracefully in 46 milliseconds
ELG shutdown gracefully in 45 milliseconds
ELG shutdown gracefully in 42 milliseconds
ELG shutdown gracefully in 83 milliseconds
ELG shutdown gracefully in 392 milliseconds
ELG shutdown gracefully in 495 milliseconds
ELG shutdown gracefully in 512 milliseconds
ELG shutdown gracefully in 420 milliseconds
ELG shutdown gracefully in 291 milliseconds
ELG shutdown gracefully in 508 milliseconds
ELG shutdown gracefully in 4105 milliseconds
ELG shutdown gracefully in 4092 milliseconds
ELG shutdown gracefully in 4182 milliseconds
ELG shutdown gracefully in 585 milliseconds
ELG shutdown gracefully in 2488 milliseconds
ELG shutdown gracefully in 2520 milliseconds
ELG shutdown gracefully in 2531 milliseconds
[ERROR] Tests run: 12, Failures: 2, Errors: 0, Skipped: 0, Time elapsed: 920.2 s <<< FAILURE! -- in EventLoopGroupShutdownTest
[ERROR] EventLoopGroupShutdownTest.shutdown(IoHandlerFactory, Class, int, int)[1] -- Time elapsed: 56.40 s <<< FAILURE!
org.opentest4j.AssertionFailedError: took longer than 10 seconds to close all client ELGs ==> expected: <true> but was: <false>
        at org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at org.junit.jupiter.api.AssertionFailureBuilder.buildAndThrow(AssertionFailureBuilder.java:132)
        at org.junit.jupiter.api.AssertTrue.failNotTrue(AssertTrue.java:63)
        at org.junit.jupiter.api.AssertTrue.assertTrue(AssertTrue.java:36)
        at org.junit.jupiter.api.Assertions.assertTrue(Assertions.java:214)
        at EventLoopGroupShutdownTest.shutdown(EventLoopGroupShutdownTest.java:57)
        at java.base/java.lang.reflect.Method.invoke(Method.java:565)
        at java.base/java.util.Optional.ifPresent(Optional.java:178)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:186)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:186)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:186)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
        at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:803)
        at java.base/java.util.stream.ReferencePipeline$7$1FlatMap.accept(ReferencePipeline.java:293)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:153)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:176)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:632)
        at java.base/java.util.stream.ReferencePipeline$7$1FlatMap.accept(ReferencePipeline.java:293)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1716)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:153)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:176)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:632)
        at java.base/java.util.stream.ReferencePipeline$7$1FlatMap.accept(ReferencePipeline.java:293)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:214)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1716)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
        at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:153)
        at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:176)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:265)
        at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:632)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)

❯ sdk current java
Alias tip: sdkc java

Using java version 25.0.1-tem
❯ java --version
openjdk 25.0.1 2025-10-21 LTS
OpenJDK Runtime Environment Temurin-25.0.1+8 (build 25.0.1+8-LTS)
OpenJDK 64-Bit Server VM Temurin-25.0.1+8 (build 25.0.1+8-LTS, mixed mode, sharing)
```

Temurin 21
```
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running EventLoopGroupShutdownTest
ELG shutdown gracefully in 95 milliseconds
ELG shutdown gracefully in 9 milliseconds
ELG shutdown gracefully in 95 milliseconds
ELG shutdown gracefully in 80 milliseconds
ELG shutdown gracefully in 77 milliseconds
ELG shutdown gracefully in 13 milliseconds
ELG shutdown gracefully in 12 milliseconds
ELG shutdown gracefully in 11 milliseconds
ELG shutdown gracefully in 1 milliseconds
ELG shutdown gracefully in 1 milliseconds
ELG shutdown gracefully in 0 milliseconds
ELG shutdown gracefully in 1 milliseconds
ELG shutdown gracefully in 1 milliseconds
ELG shutdown gracefully in 2 milliseconds
ELG shutdown gracefully in 3 milliseconds
ELG shutdown gracefully in 7 milliseconds
ELG shutdown gracefully in 8 milliseconds
ELG shutdown gracefully in 4 milliseconds
ELG shutdown gracefully in 3 milliseconds
ELG shutdown gracefully in 3 milliseconds
ELG shutdown gracefully in 6 milliseconds
ELG shutdown gracefully in 3 milliseconds
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 1.225 s -- in EventLoopGroupShutdownTest
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.707 s
[INFO] Finished at: 2026-01-23T16:28:56+13:00
[INFO] ------------------------------------------------------------------------
❯ java --version
openjdk 21.0.9 2025-10-21 LTS
OpenJDK Runtime Environment Temurin-21.0.9+10 (build 21.0.9+10-LTS)
OpenJDK 64-Bit Server VM Temurin-21.0.9+10 (build 21.0.9+10-LTS, mixed mode, sharing)
```

--- a/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemory.java
+++ b/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemory.java
@@ -19,11 +19,15 @@ import io.netty.buffer.ByteBuf;
 import io.netty.channel.socket.DatagramPacket;
 import io.netty.channel.unix.Buffer;
 import io.netty.util.internal.CleanableDirectBuffer;
+import io.netty.util.internal.PlatformDependent;
 
 import java.net.InetSocketAddress;
 import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
 
 final class MsgHdrMemory {
+    public static final int MSG_HDR_SIZE =
+            Native.SIZEOF_MSGHDR + Native.SIZEOF_SOCKADDR_STORAGE + Native.SIZEOF_IOVEC + Native.CMSG_SPACE;
     private static final byte[] EMPTY_SOCKADDR_STORAGE = new byte[Native.SIZEOF_SOCKADDR_STORAGE];
     // It is not possible to have a zero length buffer in sendFd,
     // so we use a 1 byte buffer here.
@@ -43,17 +47,30 @@ final class MsgHdrMemory {
     private final short idx;
     private final int cmsgDataOffset;
 
-    MsgHdrMemory(short idx) {
+    MsgHdrMemory(short idx, ByteBuffer msgHdrMemoryArray) {
         this.idx = idx;
-        msgHdrMemoryCleanable = Buffer.allocateDirectBufferWithNativeOrder(Native.SIZEOF_MSGHDR);
-        socketAddrMemoryCleanable = Buffer.allocateDirectBufferWithNativeOrder(Native.SIZEOF_SOCKADDR_STORAGE);
-        iovMemoryCleanable = Buffer.allocateDirectBufferWithNativeOrder(Native.SIZEOF_IOVEC);
-        cmsgDataMemoryCleanable = Buffer.allocateDirectBufferWithNativeOrder(Native.CMSG_SPACE);
-
-        msgHdrMemory = msgHdrMemoryCleanable.buffer();
-        socketAddrMemory = socketAddrMemoryCleanable.buffer();
-        iovMemory = iovMemoryCleanable.buffer();
-        cmsgDataMemory = cmsgDataMemoryCleanable.buffer();
+        this.msgHdrMemoryCleanable = null;
+        this.socketAddrMemoryCleanable = null;
+        this.iovMemoryCleanable = null;
+        this.cmsgDataMemoryCleanable = null;
+        int offset = idx * MSG_HDR_SIZE;
+        // ByteBuffer.slice(int, int) / duplicate() are specified to produce BIG_ENDIAN byte buffers.
+        // Set native order explicitly so native structs written via putInt/putLong use the expected endianness.
+        this.msgHdrMemory = PlatformDependent.offsetSlice(
+                msgHdrMemoryArray, offset, Native.SIZEOF_MSGHDR
+        ).order(ByteOrder.nativeOrder());
+        offset += Native.SIZEOF_MSGHDR;
+        this.socketAddrMemory = PlatformDependent.offsetSlice(
+                msgHdrMemoryArray, offset, Native.SIZEOF_SOCKADDR_STORAGE
+        ).order(ByteOrder.nativeOrder());
+        offset += Native.SIZEOF_SOCKADDR_STORAGE;
+        this.iovMemory = PlatformDependent.offsetSlice(
+                msgHdrMemoryArray, offset, Native.SIZEOF_IOVEC
+        ).order(ByteOrder.nativeOrder());
+        offset += Native.SIZEOF_IOVEC;
+        this.cmsgDataMemory = PlatformDependent.offsetSlice(
+                msgHdrMemoryArray, offset, Native.CMSG_SPACE
+        ).order(ByteOrder.nativeOrder());
 
         msgHdrMemoryAddress = Buffer.memoryAddress(msgHdrMemory);
 
@@ -158,11 +175,17 @@ final class MsgHdrMemory {
     }
 
     void release() {
-        msgHdrMemoryCleanable.clean();
+        if (msgHdrMemoryCleanable != null) {
+            msgHdrMemoryCleanable.clean();
+        }
         if (socketAddrMemoryCleanable != null) {
             socketAddrMemoryCleanable.clean();
         }
-        iovMemoryCleanable.clean();
-        cmsgDataMemoryCleanable.clean();
+        if (iovMemoryCleanable != null) {
+            iovMemoryCleanable.clean();
+        }
+        if (cmsgDataMemoryCleanable != null) {
+            cmsgDataMemoryCleanable.clean();
+        }
     }
 }

--- a/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
+++ b/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
@@ -88,7 +88,7 @@ final class MsgHdrMemoryArray {
         for (MsgHdrMemory hdr: hdrs) {
             hdr.release();
         }
-        this.msgHdrMemoryArrayMemoryCleanable.clean();
+        msgHdrMemoryArrayMemoryCleanable.clean();
     }
 
     int capacity() {

--- a/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
+++ b/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
@@ -85,6 +85,9 @@ final class MsgHdrMemoryArray {
     void release() {
         assert !released;
         released = true;
+        for (MsgHdrMemory hdr: hdrs) {
+            hdr.release();
+        }
         this.msgHdrMemoryArrayMemoryCleanable.clean();
     }
 

--- a/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
+++ b/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
@@ -27,9 +27,9 @@ final class MsgHdrMemoryArray {
     private final MsgHdrMemory[] hdrs;
     private final int capacity;
     private final long[] ids;
+    private final CleanableDirectBuffer msgHdrMemoryArrayMemoryCleanable;
     private boolean released;
     private int idx;
-    private final CleanableDirectBuffer msgHdrMemoryArrayMemoryCleanable;
 
     MsgHdrMemoryArray(short capacity) {
         assert capacity >= 0;

--- a/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
+++ b/transport-classes-io_uring/src/main/java/io/netty/channel/uring/MsgHdrMemoryArray.java
@@ -15,6 +15,10 @@
  */
 package io.netty.channel.uring;
 
+import io.netty.channel.unix.Buffer;
+import io.netty.util.internal.CleanableDirectBuffer;
+
+import java.nio.ByteBuffer;
 import java.util.Arrays;
 
 final class MsgHdrMemoryArray {
@@ -25,14 +29,18 @@ final class MsgHdrMemoryArray {
     private final long[] ids;
     private boolean released;
     private int idx;
+    private final CleanableDirectBuffer msgHdrMemoryArrayMemoryCleanable;
 
     MsgHdrMemoryArray(short capacity) {
         assert capacity >= 0;
         this.capacity = capacity;
         hdrs = new MsgHdrMemory[capacity];
         ids = new long[capacity];
+        int total = MsgHdrMemory.MSG_HDR_SIZE * capacity;
+        this.msgHdrMemoryArrayMemoryCleanable = Buffer.allocateDirectBufferWithNativeOrder(total);
+        ByteBuffer msgHdrMemoryArrayMemory = msgHdrMemoryArrayMemoryCleanable.buffer();
         for (int i = 0; i < hdrs.length; i++) {
-            hdrs[i] = new MsgHdrMemory((short) i);
+            hdrs[i] = new MsgHdrMemory((short) i, msgHdrMemoryArrayMemory);
             ids[i] = NO_ID;
         }
     }
@@ -77,9 +85,7 @@ final class MsgHdrMemoryArray {
     void release() {
         assert !released;
         released = true;
-        for (MsgHdrMemory hdr: hdrs) {
-            hdr.release();
-        }
+        this.msgHdrMemoryArrayMemoryCleanable.clean();
     }
 
     int capacity() {

